File: starter/python/rulebound_loader.py
from __future__ import annotations

import json
from dataclasses import dataclass
from pathlib import Path
from typing import Any


@dataclass(frozen=True)
class AssetPack:
    catalog: list[dict[str, Any]]
    finishes: list[dict[str, Any]]
    rules: dict[str, Any]
    rooms: list[dict[str, Any]]
    briefs: dict[str, str]
    historical_jobs: list[dict[str, Any]]


def read_json(path: Path) -> Any:
    return json.loads(path.read_text(encoding="utf-8"))


def load_asset_pack(input_dir: str | Path) -> AssetPack:
    root = Path(input_dir)
    rooms = [read_json(path) for path in sorted((root / "rooms").glob("*.json"))]
    briefs = {path.stem: path.read_text(encoding="utf-8").strip() for path in sorted((root / "briefs").glob("*.txt"))}
    return AssetPack(
        catalog=read_json(root / "catalog.json"),
        finishes=read_json(root / "finishes.json"),
        rules=read_json(root / "rules.json"),
        rooms=rooms,
        briefs=briefs,
        historical_jobs=read_json(root / "historical_jobs.json"),
    )

File: starter/typescript/rulebound-loader.ts
import { readFileSync, readdirSync } from "node:fs";
import { join } from "node:path";

export type JsonObject = Record<string, unknown>;
export interface AssetPack {
  catalog: JsonObject[];
  finishes: JsonObject[];
  rules: JsonObject;
  rooms: JsonObject[];
  briefs: Record<string, string>;
  historicalJobs: JsonObject[];
}

const readJson = <T>(path: string): T => JSON.parse(readFileSync(path, "utf8")) as T;

export function loadAssetPack(inputDir: string): AssetPack {
  const roomDir = join(inputDir, "rooms");
  const briefDir = join(inputDir, "briefs");
  const roomFiles = readdirSync(roomDir).filter((name) => name.endsWith(".json")).sort();
  const briefFiles = readdirSync(briefDir).filter((name) => name.endsWith(".txt")).sort();
  return {
    catalog: readJson<JsonObject[]>(join(inputDir, "catalog.json")),
    finishes: readJson<JsonObject[]>(join(inputDir, "finishes.json")),
    rules: readJson<JsonObject>(join(inputDir, "rules.json")),
    rooms: roomFiles.map((name) => readJson<JsonObject>(join(roomDir, name))),
    briefs: Object.fromEntries(briefFiles.map((name) => [name.replace(/\.txt$/, ""), readFileSync(join(briefDir, name), "utf8").trim()])),
    historicalJobs: readJson<JsonObject[]>(join(inputDir, "historical_jobs.json")),
  };
}
